src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs

```diff
@@ -235,70 +244,358 @@ internal static void FailFast(string message, Exception? exception, RhFailFastRe
+                goto error;
+            if (!callback())
+                goto error;
+            if (!CloseValue(closing))
+                goto error;
+            return true;
+        error:
+            _currentBufferIndex = savedIndex;
+            return false;
+        }
+
+        /// <summary>
+        /// Write raw bytes or already converted to UTF8 string to the triage buffer. If the block can not
+        /// fit in the triage buffer the allocations are backed out and the value is skipped.
+        /// </summary>
+        /// <param name="key">the name of the value</param>
+        /// <param name="bytes">value</param>
+        /// <returns>true - success, false - out of triage buffer space</returns>
+        private static bool WriteValue(ReadOnlySpan<char> key, ReadOnlySpan<byte> bytes)
+        {
+            int savedIndex = _currentBufferIndex;
+            if (!OpenValue(key, '"'))
+                goto error;
+            if (!WriteSpan(bytes))
+                goto error;
+            if (!CloseValue('"'))
+                goto error;
+            return true;
+        error:
+            _currentBufferIndex = savedIndex;
+            return false;
+        }
+
+        private static bool OpenValue(ReadOnlySpan<char> key, char marker)
+        {
+            if (!WriteSeparator())
+                return false;
+            if (!key.IsEmpty)
+            {
+                if (!WriteChar('"'))
+                    return false;
+                if (!WriteChars(key))
+                    return false;
+                if (!WriteChar('"'))
+                    return false;
+                if (!WriteChar(':'))
+                    return false;
+            }
+            if (!WriteChar(marker))
+                return false;
+            _comma = false;
+            return true;
+        }
+
+        private static bool CloseValue(char marker)
+        {
+            if (!WriteChar(marker))
+                return false;
+            _comma = true;
+            return true;
+        }
+
+        private static bool WriteSeparator() => _comma ? WriteChar(',') : true;
+
+        private static bool WriteChar(char source) => WriteChars(new ReadOnlySpan<char>(source));
+
+        private static bool WriteChars(ReadOnlySpan<char> chars, int max = int.MaxValue)
+        {
+            int size = Encoding.UTF8.GetByteCount(chars);
+            size = Math.Min(size, max);
+            Span<byte> destination = AllocBuffer(size);
+            if (destination.IsEmpty)
+            {
+                return false;
+            }
+            Encoding.UTF8.GetBytes(chars.Slice(0, size), destination);
+            return true;
+        }
+
+        private static bool WriteSpan(ReadOnlySpan<byte> bytes)
+        {
+            Span<byte> destination = AllocBuffer(bytes.Length);
+            if (destination.IsEmpty)
+            {
+                return false;
+            }
+            bytes.CopyTo(destination);
```

Comment:
Does this need escaping of special characters to make it a valid json?

Reply:
Yes, I need to add that.